source/common/upstream/load_balancer_impl.cc

```diff
@@ -337,47 +337,146 @@ const HostVector& ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts(const H
   const HostSet& host_set = chooseHostSet();
+  HostsSource hosts_source;
+  hosts_source.priority_ = host_set.priority();
 
   // If the selected host set has insufficient healthy hosts, return all hosts.
   if (isGlobalPanic(host_set)) {
     stats_.lb_healthy_panic_.inc();
-    return host_set.hosts();
+    hosts_source.source_type_ = HostsSource::SourceType::AllHosts;
+    return hosts_source;
   }
 
   // If we've latched that we can't do priority-based routing, return healthy hosts for the selected
   // host set.
   if (per_priority_state_[host_set.priority()]->locality_routing_state_ ==
       LocalityRoutingState::NoLocalityRouting) {
-    return host_set.healthyHosts();
+    hosts_source.source_type_ = HostsSource::SourceType::HealthyHosts;
+    return hosts_source;
   }
 
   // Determine if the load balancer should do zone based routing for this pick.
   if (!runtime_.snapshot().featureEnabled(RuntimeZoneEnabled, 100)) {
-    return host_set.healthyHosts();
+    hosts_source.source_type_ = HostsSource::SourceType::HealthyHosts;
+    return hosts_source;
   }
 
   if (isGlobalPanic(localHostSet())) {
     stats_.lb_local_cluster_not_ok_.inc();
     // If the local Envoy instances are in global panic, do not do locality
     // based routing.
+    hosts_source.source_type_ = HostsSource::SourceType::HealthyHosts;
+    return hosts_source;
+  }
+
+  hosts_source.source_type_ = HostsSource::SourceType::LocalityHealthyHosts;
+  hosts_source.locality_index_ = tryChooseLocalLocalityHosts(host_set);
+  return hosts_source;
+}
+
+const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts_source) {
+  const HostSet& host_set = *priority_set_.hostSetsPerPriority()[hosts_source.priority_];
+  switch (hosts_source.source_type_) {
+  case HostsSource::SourceType::AllHosts:
+    return host_set.hosts();
+  case HostsSource::SourceType::HealthyHosts:
     return host_set.healthyHosts();
+  case HostsSource::SourceType::LocalityHealthyHosts:
+    return host_set.healthyHostsPerLocality().get()[hosts_source.locality_index_];
+  default:
+    NOT_REACHED;
   }
+}
 
-  return tryChooseLocalLocalityHosts(host_set);
+RoundRobinLoadBalancer::RoundRobinLoadBalancer(
+    const PrioritySet& priority_set, const PrioritySet* local_priority_set, ClusterStats& stats,
+    Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+    const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+    : ZoneAwareLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
+                                common_config) {
+  for (uint32_t priority = 0; priority < priority_set.hostSetsPerPriority().size(); ++priority) {
+    refresh(priority);
+  }
+  // We fully recompute the schedulers for a given host set here on membership change, which is
+  // consistent with what other LB implementations do (e.g. thread aware).
+  // TODO(htuch): By fully recomputing the host set schedulers on each membership update, we lose RR
```

Comment:
In practice I really doubt this is going to be an issue but seems worth it to leave the TODO